### PR TITLE
Feature/1 beacon api

### PR DESCRIPTION
**Pull request by eddiejaoude**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/app/bootstrap.php.cache
+/app/cache/*
+!app/cache/.gitkeep
+/app/config/parameters.yml
+/app/logs/*
```
Comment (brettminnie) [inline .gitignore]: technically app/logs/ will work fine here you don't need the trailing *

Comment (eddiejaoude) [inline .gitignore]: I think that all came from symfony. But good to know :smile:

File: app/config/services_dev.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+services:
```
Comment (eddiejaoude) [inline app/config/services_dev.yml]: @brettminnie Note: I want to move this file into the bundle

Comment (eddiejaoude) [inline app/config/services_dev.yml]: Moved

File: src/DashboardHub/Bundle/BeaconBundle/Client/HttpClient.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?php
+namespace DashboardHub\Bundle\BeaconBundle\Client;
+
+use GuzzleHttp\Client;
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Client/HttpClient.php]: maybe cast this to a better name, ie GuzzleHttpClient, just makes it more readable

File: src/DashboardHub/Bundle/BeaconBundle/Client/HttpClientInterface.php [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ */
+interface HttpClientInterface
+{
+
+    /**
+     * @param string $uri
+     *
+     * @return array
+     */
+    public function get($uri);
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Client/HttpClientInterface.php]: if you want to match HTTP verbs these should be in upper case, I have no preference, just a thought

File: src/DashboardHub/Bundle/BeaconBundle/Controller/DefaultController.php [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<?php
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Controller/DefaultController.php]: Do you need the default symfony controller?

File: src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                $this->get('dashboard_hub_beacon.dashboard.api')
+                     ->get('/api/dashboards/' . $uid);
+        } catch (\Exception $e) {
+            return $this->handleView(
+                $this->view(new \stdClass(), 404)
+            );
+        }
+
+        $page = $this->get('dashboard_hub_beacon.service.yslow')
+                     ->save($uid, $tag, $this->get('request')->request->all());
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php]: personally, I would write this line as 
```php 
->save(
    $uid,
    $tag,
    $this->get...
);
```

As the last argument is quite chained.

File: src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     * @var int
+     */
+    protected $requests = 0;
+
+    /**
+     * @var int
+     */
+    protected $pageLoadTime = 0;
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php]: :+1:  in general for you initializing all your variables with sensible defaults

Turn you into a real programmer still :D

File: src/DashboardHub/Bundle/BeaconBundle/Client/HttpClientInterface.php (hunk 4 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Client/HttpClientInterface.php]: I was matching to what `Guzzle` used. 

Not sure if i like the idea of making it upper case, but a good idea.

File: src/DashboardHub/Bundle/BeaconBundle/Client/HttpClient.php (hunk 3 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Client/HttpClient.php]: Good idea.

File: src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     * @var array
+     */
+    protected $raw = array();
+
+    /**
+     * @var \Datetime
+     */
+    protected $createdOn;
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php]: you have initialized everything else, put this in a constructor with `new \DateTime()`

File: .scrutinizer.yml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  php_pdepend: true
+  php_analyzer: true
+checks:
+    php:
+        code_rating: true
+        duplication: true
+
+filter:
+    paths: [src/*]
+    excluded_paths: [Tests/*, test/*, src/DashboardHub/BeaconBundle/Tests/*]
```
Comment (jeremyquinton) [inline .scrutinizer.yml]: why no scrutinizer on tests testing code is as important as the actual code?

File: src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php (hunk 6 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php]: Yep good idea!

File: src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php (hunk 7 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php]: :wink:

File: src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php (hunk 8 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Entity/Page.php]: :+1:

File: .scrutinizer.yml (hunk 9 above)
Comment (eddiejaoude) [inline .scrutinizer.yml]: True! Lets do it!!

File: src/DashboardHub/Bundle/BeaconBundle/Features/yslow-beacon.feature [hunk 10]
```diff
@@ -0,0 +1,35 @@
+Feature: YSlow Beacon
+
+  Scenario: Successful save beaconed data
+    Given I called "PostYSlowBeacon" with the following values from JSON:
+    """
+    {
+        "uid": "123abc",
+        "tag": "tag-abc",
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Features/yslow-beacon.feature]: Up until this point you have named everything quite sanely, now we get this down with the kids abbreviations, you don't remember what you had for breakfast, I'm as bad, there is no way that in a month, looking at a json body we are going to remember what this all means

Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Features/yslow-beacon.feature]: Up until this point you have named everything quite sanely, now we get this down with the kids abbreviations, you don't remember what you had for breakfast, I'm as bad, there is no way that in a month, looking at a json body we are going to remember what this all means

File: src/DashboardHub/Bundle/BeaconBundle/Resources/config/services_dev.yml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+services:
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Resources/config/services_dev.yml]: :+1:

File: src/DashboardHub/Bundle/BeaconBundle/Service/YslowService.php [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+     *
+     * @return int
+     */
+    public function save($uid, $tag, array $yslow)
+    {
+        $page = new Page();
+        $page->setDashboardUid($uid);
+        $page->setCreatedOn(new \Datetime());
+        $page->setTag($tag);
+        $page->setPageLoadTime($yslow['lt']);
```
Comment (brettminnie) [inline src/DashboardHub/Bundle/BeaconBundle/Service/YslowService.php]: see my comments about about this data naming

**Comment by brettminnie** [comment]
As a general suggestion, possibly do the framework setup without any customisation on dev, that way we don't have to look at all the symfony core in the pull request. Will make the initial PR a bit less painful

File: src/DashboardHub/Bundle/BeaconBundle/Features/yslow-beacon.feature (hunk 10 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Features/yslow-beacon.feature]: Are you referring to the single letter abbreviations? I am not keen on such abbreviations too (ignoring that I used it on the urls for the dashboard). 

However, these...
```javascript
...
"w": 786226,
    "o": 87,
    "u": "http%3A%2F%2Flocalhost%3A8000%2F",
    "r": 11,
    "lt": 1100,
    "i": "ydefault",
    "g": {
...
```
 are from `yslow`, so they can't be changed :cry:

File: .scrutinizer.yml (hunk 9 above)
Comment (brettminnie) [inline .scrutinizer.yml]: Worth a go

**Comment by eddiejaoude** [comment]
> As a general suggestion, possibly do the framework setup without any customisation on dev, that way we don't have to look at all the symfony core in the pull request. Will make the initial PR a bit less painful

Good point!

File: src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php (hunk 6 above)
Comment (jeremyquinton) [inline src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php]: Another idea would be if you took ``` $this->get('request')->request->all() ``` and assigned it to a variable and then passed it to the function so 

```
$descriptiveVariableName = $this->get('request')->request->all();
$page = $this->get('dashboard_hub_beacon.service.yslow')
                      ->save($uid, $tag, $descriptiveVariableName);
```

File: src/DashboardHub/Bundle/BeaconBundle/DataFixtures/ORM/LoadYslowData.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+use Doctrine\Common\Persistence\ObjectManager;
+
+class LoadYslowData implements FixtureInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function load(ObjectManager $manager)
+    {
+        for($i=1;$i<=10;$i++) {
```
Comment (jeremyquinton) [inline src/DashboardHub/Bundle/BeaconBundle/DataFixtures/ORM/LoadYslowData.php]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#54-for

File: src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php (hunk 6 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/Controller/YslowController.php]: :+1:

File: src/DashboardHub/Bundle/BeaconBundle/DataFixtures/ORM/LoadYslowData.php (hunk 13 above)
Comment (eddiejaoude) [inline src/DashboardHub/Bundle/BeaconBundle/DataFixtures/ORM/LoadYslowData.php]: :+1:

**Pull request merged**